### DISCUSSION TITLE
[HELP] Import requirements from CSV/EXCEL to CAPELLA

### DISCUSSION THREAD

**Discussion by jeremy-jimmy** [dbinfrago/py-capellambse#105] [Q&A] [answered]
Hi, 
I'm currently trying to build a python function able to operate export/import of the requirements stocked in my Capella to an EXCEL file.

Using the package "capellambse" I manage quite easily to handle the export from CAPELLA to CSV/EXCEL but I'm facing difficulties to handle the other way.

My main problem is that so far I didn't acheive to custom attributes to my requirement "on the fly" with my script.

`
model = capellambse.MelodyModel(PATH_CAPELLA, jupyter_untrusted=True)
REQ = model.sa.requirements.create(name="TEST_XTYPE", 
                                                            xtype="Requirements:Requirement")
`

How to add custom attributes fields to my object REQ ?

Thanks in advance for support,

Jérémy

**Comment by vik378** [comment]
nice use case indeed, I think we can add an example for how to create a req attribute into https://github.com/DSD-DBS/py-capellambse/blob/master/examples/06%20Introduction%20to%20Requirement%20access%20and%20management.ipynb or maybe even an example for excel import with attribute definition creation. The catch there was about having the definitions / enums pre defined or imported upfront. @ewuerger , @Wuestengecko - since you are deep in this topic now,  would any of you have time to add a quick example for attr instantiation?

**Reply by jeremy-jimmy** [reply]
That would be great to add in the jupyter notebooks indeed, thanks in advance

**Comment by ewuerger** [comment, marked as answer]
> How to add custom attributes fields to my object REQ ?

Capella (Requirements plugin) is doing that in an overengineered way. While implementing some Requirement Bots that automate maintenance we used the API to its full extent already. Just to get you going instead of waiting for a demo notebook:
In order to add `AttributeValue`s to a Requirement you can use the same create functionality as seen in the notebook. Here it would go like this:

```python
from capellambse.extensions import reqif

requirement = model.search(reqif.XT_REQUIREMENT)[0]
requirement.attributes.create("int", name="Example", value=10)
```
If you forget to handover the type hint (here "int") then the API will tell you to do so, since this is needed for creation. Possible types are:
- int, integer, integervalueattribute
- str, string, stringvalueattribute
- float, real, realvalueattribute
- date, datevalueattribute
- bool, boolvalueattribute
- enum, enumvalueattribute

Then this is all you need to create attribute values underneath your requirements. But if you want to add a requirement type and/or definition to the `AttributeValue` you'll need to make sure that a `RequirementTypesFolder` exists (can also be created via API) with a `RequirementType` and `AttributeDefinition`s underneath. It is your choice then to set these attributes on your new requirement via:

```python

reqtfolder = model.la.requirement_types_folders[0]
reqtype = reqtfolder.requirement_types[0]
reqtype.attribute_definitions.create(reqif.XT_REQ_TYPE_ATTR_DEF, long_name="Example Definition")
new_req_attr = requirement.attributes[-1]
new_req_attr.type = reqtype
new_req_attr.definition = reqtype.attribute_definitions[0]
```

or handover while creation of attributes:

```python

definition = reqtype.attribute_definitions.by_long_name("Example Definition")
requirement.attributes.create("int", name="Example", type=reqtype, definition=definition)
```

We like the latter more. Also this will improve the display of `AttributeValue`s inside of Capella (since the name of the definition is displayed).
Personally I think that normal `AttributeDataTypeDefinition` don't have a use for now. But underneath `EnumAttributeDataTypeDefinition` you can define the `EnumValue`s. These will be available for `EnumAttributeValue`s if you set the data_type attribute on the `EnumAttributeDefinition` correctly. Keep that in mind.

Key takeaways:
- Creation is enabled on attributes that handback an `ElementList`
- When creating elements where multiple classtypes are possible the API will raise a `ValueError` to inform you. You'll need to find the right XTYPE string in the reqif module. We'll work on refactoring this to also take class-types right away.

We'll work on improving documentation and showcases in the future.

After that it will be a walk in the park to get your Pandas DataFrame content back into Capella. Maybe if you are invested, you can try to extent reqif to offer import from excel-file?

**Reply by jeremy-jimmy** [reply]
Thank for the fast answer, I will read this carefully and I hope I will manage to handle my problem with your help.

**Reply by jeremy-jimmy** [reply]
I managed to handle my use-case with your help, thanks again

**Reply by Bhuertas** [reply]
Hi,

I have an issue while trying your first method 

> requirement.attributes.create("int", name="Example", value=10)

When I try to create an attribute I get an error saying that we cannot set name on a stringvalueattribute

The error :  
```
Traceback (most recent call last):
  File "C:\Users\bahuerta\Capella\capella\workspace\Python4Capella\VSC_Python_scripts\xslx2capella.py", line 52, in <module>
    new_req.attributes.create("str",name = attributs[cell.column-1], value = cell.value)
  File "C:\Users\bahuerta\AppData\Local\Programs\Python\Python311\Lib\site-packages\capellambse\model\_obj.py", line 2029, in create
    return self._insert(len(self), marker)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bahuerta\AppData\Local\Programs\Python\Python311\Lib\site-packages\capellambse\model\_obj.py", line 2074, in _insert
    value = acc.insert(self, index, value)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bahuerta\AppData\Local\Programs\Python\Python311\Lib\site-packages\capellambse\model\_descriptors.py", line 3011, in insert
    value = self._insert_create(
            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bahuerta\AppData\Local\Programs\Python\Python311\Lib\site-packages\capellambse\model\_descriptors.py", line 3133, in _insert_create
    return cls(
           ^^^^
  File "C:\Users\bahuerta\AppData\Local\Programs\Python\Python311\Lib\site-packages\capellambse\model\_obj.py", line 746, in __init__
    raise TypeError(
TypeError: Cannot set 'name' on StringValueAttribute
```

If someone has any idea of how I can solve that, thanks in advance !

**Reply by ewuerger** [reply]
That is true! You can't set a name or type on Requirement attributes anymore, but you can't do that in the Capella 7.0.0 UI either. Now what is only left to set is:
<img width="987" height="185" alt="image" src="https://github.com/user-attachments/assets/5899a2fa-d40c-4a88-aca5-49d93a8bbac5" />

`definition` and `value`. The metamodel is generated from the source code of Capella such that capellambse covers full accessability for any data. Therefore the snippet I provided can be reduced to just:
```
requirement.attributes.create("int", definition=definition, value=10)
```

**Comment by Luffy-cpu** [comment]
Hi,

I have a question related to this one. In the case the attribute already exists in the model, if I want to update its value, how can I do it? Which method should I call, please?

Thanks in advance,
Christian

**Reply by vik378** [reply]
I think we'll need a bit more context (at least an example) about what you are trying to update / attribute of what. If it is about Requirement object I have a feeling that you may find many interesting things here: https://github.com/DSD-DBS/capella-rm-bridge - this was our early attempt at mass-syncing requirements (import and update on change) from a zoo of RM tools into Capella.

**Reply by Luffy-cpu** [reply]
Sorry, I didn't give precision.

I have a dict in python that contains the new values of requirements attributes, for example 
req = {"Id": "R1", "Text": "The requirement shall be complete", "Rationale": "From Specification Doc"}
I assume that in the Capella model, the corresponding attributes are "identifier", "text" and "rationale" (the last one is a custom attribute).
I want to update the text and rationale attributes of the requirement which identifier is equal to "R1" in the Capella with my dict.
I thought that a setter function should exist to update the values, but I didn't find it.

Thank you for the RM-Bridge project. I will install it.

**Reply by Luffy-cpu** [reply]
Hi again,
I installed capella-rm-bridge, but I'm not sure I understand how to use it to update requirements attributes in Capella in the context I described above. Can you help me with this, please?

**Reply by Wuestengecko** [reply]
Since you already have all the info in Python dicts, it might be easiest to use the [declarative YAML](https://dsd-dbs.github.io/py-capellambse/start/declarative.html) facility here. It has the "sync" command that will update existing attributes (if they exist already) or create new ones in the correct location (if they don't).

In your specific case, starting with a list of dicts like this:

```python
my_requirements = [
    {
        "Id": "R1",
        "Text": "The requirement shall be complete",
        "Rationale": "From Specification Doc"
    },
    {
        "Id": ..., # etc.
    },
    ...
]
```

I'm assuming the "Id" here is used to find a Requirement based on its [`identifier`](https://dsd-dbs.github.io/py-capellambse/code/capellambse.extensions.reqif.html#capellambse.extensions.reqif.ReqIFElement.identifier), the "Text" is put into the [`text`](https://dsd-dbs.github.io/py-capellambse/code/capellambse.extensions.reqif.html#capellambse.extensions.reqif.Requirement.text) field, and the "Rationale" should update or create a String attribute named "Rationale". If these assumptions are not correct, you'll have to modify the code a little :)

The basic structure of the YAML you need would be:

```yaml
- parent: !find
    _type: Requirement
    identifier: <Id>
  set:
    text: <Text>
  sync:
    attributes:
      - find:
          _type: StringValueAttribute  # unfortunately this doesn't support the short names currently
          name: Rationale
        set:
          value: <Rationale>
- parent: !find ... # etc.
- parent: !find ...
```

Using the helpers from `capellambse.decl`, use can do that like this:

```python
# Build the YAML
from capellambse import decl
instructions = []
for req in my_requirements:
    instructions.append({
        "parent": decl.FindBy({
            "_type": "Requirement",
            "identifier": req["Id"],
        }),
        "set": {"text": req["Text"]},
        "sync": {
            "attributes": [{
                "find": {
                    "_type": "StringValueAttribute",
                    "name": "Rationale"
                },
                "set": {"value": req["Rationale"]},
            }]
        }
    })
yaml = decl.dump(instructions)

# Apply it to the model
import capellambse, io
model = capellambse.MelodyModel(...)
decl.apply(model, io.StringIO(yaml))
model.save()
```

If you have any further questions, don't hesitate to ask. :)

**Reply by Luffy-cpu** [reply]
Thanks a lot for the answer.
My use case is exactly what you understood.
I tried as explained, but just ran into the following AttributeError:
"AttributeError: module 'capellambse.decl' has no attribute 'FindBy'"
I tried to update my capellambse and capella-rm-bridge packages but it did not help me fixing the problem.

**Reply by Wuestengecko** [reply]
> AttributeError: module 'capellambse.decl' has no attribute 'FindBy'

That's weird, FindBy has been around since end of March. Can you please post the output of these commands?

    python --version

    python -c "import capellambse; print(capellambse.__version__, capellambse.__file__)"

    pip show capellambse

**Reply by Luffy-cpu** [reply]
Here the result for the commands

 python --version: Python 3.9.13

python -c "import capellambse; print(capellambse.__version__, capellambse.__file__)": 0.5.41 path_to_python/python3.9/site-packages/capellambse/__init__.py

pip show capellambse:
Name: capellambse
Version: 0.5.41
Summary: Provides access to Capella MBSE projects in Python
Home-page: 
Author: DB Netz AG
Author-email: 
License: Apache-2.0
Location: path_to_python/python3.9/site-packages
Requires: lxml, markupsafe, Pillow, platformdirs, svgwrite, typing-extensions
Required-by: capella-RM-Bridge

**Reply by Wuestengecko** [reply]
Ah, that's what I suspected ... you're using Python 3.9. Newer versions of capellambse use some 3.10+ specific features, so you'll have to [upgrade to a newer version](https://www.python.org/downloads/) in order to use those. If it's an option for you, I'd suggest to go straight to the newest stable version 3.12, but 3.10 should also be fine for the time being.

**Reply by Luffy-cpu** [reply]
Okay, I understand. I upgraded to Python 3.12 and decl.FindBy works correctly now and I am able to generate the right YAML.

While saving the model, the following error occurs: 
CorruptModelError(
capellambse.loader.core.CorruptModelError: Refusing to save a corrupt model without having a backup (hint: pass i_have_a_recent_backup=True)

So I tried with "model.save(i_have_a_recent_backup=True)". The error message no longer appears. But is it the recommended way to do it?

Finally, the text attribute is updated in my Capella model, but the custom attribute Rationale remains unchanged. Is there something I did wrong?

**Reply by Wuestengecko** [reply]
This error is a safeguard against making things even worse if you had to load the model with `ignore_duplicate_uuids_and_void_all_warranties=True`. The recommended way is to try and get rid of that :)

The duplication check used to be too strict in older versions of capellambse (in some cases, duplication is harmless), so now that you're up to date with latest stable, you might not even need that anymore anyways.

As for why the custom attribute is missing, I don't know yet, I'll try if I can reproduce that.

**Reply by Luffy-cpu** [reply]
Thank you very much.
I understand better about he safeguard.

For the custom attribute issue, I managed to import it, but I still have a bug. I have  now 2 attributes for Rationale after the import.
As shown in the following screenshot, a new attribute called [String Value Attribute] is created with the value of the Rationale attribute.  And the initial Rationale attribute remains with its old value.

![image](https://github.com/DSD-DBS/py-capellambse/assets/74272966/8703e160-bf93-4df0-b6ec-2ac65a883633)

I have one more question : in addition to attribute update, if I want to update requirement relations, how can I perfom it, please ?

Thank you again
